[ENH] Faster and more flexible code, and code sharing for kernel tests

adam2392 opened the pull request:
<!-- To ensure we can review your pull request promptly please complete this template entirely. -->

<!-- Please reference the issue number here. You can replace "Fixes" with "Closes" if it makes more sense. -->
Towards: #15 

Changes proposed in this pull request:
<!-- Please list all changes/additions here. -->
- refactors code to setup for kcd test
- allows any of the pairwise kernel strings to be passed in from sklearn (which is significantly faster than using partial because sklearn optimizes the in-house kernels)
- also requires kernel functions to be a specific API, so it's easier to test, implement and document

This should all make implementation of the kcd test pretty straightforward

## Before submitting

<!-- Please complete this checklist BEFORE submitting your PR to speed along the review process. -->
- [ ] I've read and followed all steps in the [Making a pull request](https://github.com/py-why/pywhy-stats/blob/main/CONTRIBUTING.md#making-a-pull-request)
    section of the `CONTRIBUTING` docs.
- [ ] I've updated or added any relevant docstrings following the syntax described in the
    [Writing docstrings](https://github.com/py-why/pywhy-stats/blob/main/CONTRIBUTING.md#writing-docstrings) section of the `CONTRIBUTING` docs.
- [ ] If this PR fixes a bug, I've added a test that will fail without my fix.
- [ ] If this PR adds a new feature, I've added tests that sufficiently cover my new functionality.
- [ ] I have added a changelog entry succintly describing the change in this PR in the [whats_new](https://github.com/py-why/pywhy-stats/blob/main/docs/whats_new/) relevant version document.

## After submitting

<!-- Please complete this checklist AFTER submitting your PR to speed along the review process. -->
- [ ] All GitHub Actions jobs for my pull request have passed.

bloebp commented on pywhy_stats/api.py:
```diff
@@ -59,32 +60,37 @@ def independence_test(
 
             # XXX: we should add pinguoin as an optional dependency for doing multi-comp stuff
-            if pval < 0.05:
+            if np.atleast_1d(res.pvalue).any() < 0.05:
```
Don't get this line, are there multiple p-values? I thought the scipy test only returns one (with the test statistic).

adam2392 replied:
Yes if the columns are multivariate, then it is an array of pvalues: https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.normaltest.html

bloebp commented on pywhy_stats/api.py:
```diff
@@ -59,32 +60,37 @@ def independence_test(
 
     if condition_on is None:
-        return method_module.ind(X, Y, method, **kwargs)
+        return method_module.value.ind(X, Y, **kwargs)
```
The `value` part is very confusing, why do we need this? Or, why don't we keep the structure as before?

adam2392 replied:
Because it actually did not previously work. Since there was no test, we didn't catch it. 

Take `method_module = Enum(FISHERZ)` for example. 

`method_module.name == 'FISHERZ'` and `method_module.value == fisherz` (the function). Enumeration is kinda wonky imo because of this, but it does make things like `method_module == METHODS.FISHERZ` easier

bloebp replied:
Oh I see. Can you add a test then to make sure that all methods in this high-level function is working with the enums?

adam2392 replied:
Added in `tests/test_api.py`

bloebp commented on pywhy_stats/independence/kci.py:
```diff
@@ -153,40 +204,19 @@ def _kernel_test(
-    Kx = _fast_centering(Kx)
-    Ky = _fast_centering(Ky)
+        # concatenate the (X, Z) data to compute the K_xz kernel
+        X = np.concatenate((X, Z), axis=1)
```
This will cause error with customized kernels and/or mixed data types. For instance, if X is continous and Z is categorical (strings) and a user only defines a RBF kernel (or we use it by default), we cannot apply it to the concatenated version. Therefore, I used a pointwise multiplication of the kernel matrices before (which is equivalent), in that way, we still have our individual kernels for each variable.

bloebp replied:
In this regards, the `kernel_Z` is not used anymore?

bloebp replied:
Another follow-up: Providing these custom kernels does not seem to be compatible with the `compute_kernel` logic. What about a simple logic like:
```
if isinstance(kernel_X, str):
   kernel_X = compute_kernel(X, metric=kernel_X, centered=centered, n_jobs=n_jobs)

Kx = kernel_X(X)
```

then, we can still support strings and use the other metric logic from scikit, but can also allow completely customized kernels. The only thing that becomes slightly more complicated would custom metric implementations, but even there, a user can simply pass: `kernel_X = lambda x: compute_kernel(x, metric=my_custom_metric, centered=centered, n_jobs=n_jobs)` (or use partial).

What do you think? I feel like this is a compromise that allows both.

adam2392 replied:
Related: https://github.com/py-why/pywhy-stats/pull/19/files#r1287466282

I think I have to think about this a bit more. 

```
if isinstance(kernel_X, str):
   kernel_X = compute_kernel(X, metric=kernel_X, centered=centered, n_jobs=n_jobs)

Kx = kernel_X(X)
```

inherently assumes tho the `kernel_X` callable is a certain API signature. If we tell users in the doc-string this is how it will be called, will this be sufficient? I am okay w/ this pattern, but this pattern was not sufficient to cover all use-cases I had so hence the refactoring.

But the if/else pattern can then capture both cases. Just need to make sure this is what we **require** in the documentation, otw we just error out and say the callable they passed is incorrect?

bloebp commented on tests/test_kci.py:
```diff
@@ -1,19 +1,22 @@
 # Unconditional independence tests - Continuous
 ####################################################
+seed = 12345
+np.random.seed(seed)
```
We should not set a seed here, since will influence other downstream pytests as well. We can do it similar as in the other tests where we used a initialized `rng` object (I guess the have something like that for numpy as well?). Then, this should only be locally.

adam2392 replied:
Okay I added a RNG that is then sequentially used. Lmk if this is fine

bloebp commented on tests/test_kci.py:
```diff
@@ -1,19 +1,22 @@
+np.random.seed(seed)
 
 
 @flaky(max_runs=3)
```
If we work with a fixed see, no need for flaky decorator. Note that the fixed seed are generating consistently the same values across different OS and python versions etc.

adam2392 replied:
Is the suggestion then to remove random seed and keep flaky I suppose?

bloebp replied:
I am actually also fine with a fixed seed (as long as it doesn't impact downstream tests). However, I still see the benefit of having different generated data every run to avoid optimizing for particular set of data. For instance, 200 samples for the KCI test is already quite small and might fail with a different seed. However, is this now due to a subtle issue in the implementation or just a general statistical issue?

adam2392 replied:
I'll just remove the seed for now and we can revisit the rng vs seed vs flaky option in the unit-tests if it is an issue

bloebp commented on tests/test_kci.py:
```diff
@@ -210,16 +220,16 @@ def test_given_gaussian_data_and_polynomial_kernel_when_perform_kernel_based_tes
 
-    assert kci.ind(X, X1, kernel_X=polynomial_kernel, kernel_Y=polynomial_kernel).pvalue > 0.05
-    assert kci.ind(X, Z, kernel_X=polynomial_kernel, kernel_Y=polynomial_kernel).pvalue < 0.05
+    assert kci.ind(X, X1, kernel_X="polynomial", kernel_Y="polynomial").pvalue > 0.05
+    assert kci.ind(X, Z, kernel_X="polynomial", kernel_Y="polynomial").pvalue < 0.05
```
Just to make sure, it is still possible to provide a callable here, right?

adam2392 replied:
Yes, see https://github.com/py-why/pywhy-stats/pull/19/files#r1287466282

adam2392 replied:
Do you mean `def my_custom_kernel(X)`? It is tested above and still works, but as we discussed in the [other comment](https://github.com/py-why/pywhy-stats/pull/19/files#r1287477665), I can refactor the inner-workings.

I'll add this soon

adam2392 replied:
Done.

bloebp commented on pywhy_stats/kernel_utils.py:
```diff
@@ -0,0 +1,281 @@
+        kernel = metric(X)
+    else:
+        kernel = pairwise_kernels(
+            X, Y=Y, metric=metric, n_jobs=n_jobs, filter_params=False, **kernel_params
```
What if the `metric` is already providing a kernel matrix? This is, what if a user provides a custom kernel implementation instead of a metric? Currently, it is confusing that we ask for a `kernel_X` etc., but it is actually supposed to be a `metric` or do I miss something?

adam2392 replied:
The `if` statement above is meant to handle this and allow a user to provide a custom kernel implementation. So my take is we should have easy support for existing optimized kernels: e.g. scikit-learn/scipy strings. Therefore `kernel_X` can both be a callable that is documented [here](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.pairwise.pairwise_kernels.html), or a string that we vectorize.

I agree we should also support custom kernels, but then we want the code to be consistent. In the previous implementation, the API inherently assumed it was an input of just `X`. But then it doesn't support pairwise kernels among X and Y. I think an easy way to handle this is to require the user to define a lambda that takes in `X, Y=None` and if Y is not used, then it is not used, but then we can have a consistent API signature.

I think the current way I handled it is not ideal yet, but it's just a draft to get us to converge on something that generalizes to the rest of the codebase.

bloebp replied:
Ah ok, I thought it would need to be a metric then (not a kernel). In that way, it is basically similar to my suggestion above with the if-else. I see the point with `Y = None`, however, I would simply check if Y is None and if not, pass it as second argument. It will then either raise an error or the user provided a kernel that accepts a second argument. No need to add additional validation logic.

adam2392 replied:
I refactored it to use your suggestion. Is this what you meant?

adam2392 replied:
See lines above this. I.e.

```
    if callable(metric):
        # If the number of arguments is just one, then we bypass the pairwise kernel
        # optimized computation via sklearn and opt to use the metric function directly
        input_args = [X]
        if Y is not None:
            input_args.append(Y)
        kernel = metric(*input_args)
```

adam2392 commented:
Tldr re: https://github.com/py-why/pywhy-stats/pull/19#discussion_r1287466282

I will put up a sep commit after Neurips rebuttal period is over to see if we can converge :)

bloebp commented:
Can you investigate the failing checks?

adam2392 commented:
@bloebp lmk wdyt. I addressed the two issues:

1. use `RNG`, so the seed does not universally affect the tests
2. I refactored, but I think there might be still a discrepancy:

- on one hand `pairwise_kernels` can take a callable that is defined on (X, Y) 1D row vectors and will run a joblib parallelization
- on the other hand if you define your own kernel callable that can compute this in a vectorized fashion, then you want to do that.

E.g.
```
    def my_custom_kernel(X):
        # X here is multidimensional, where the first dimension is continuous and the second categorical. We can handle
        # this by combining two kernels.
        return delta_kernel(X[:, 0].reshape(-1, 1)) * rbf_kernel(X[:, 1].reshape(-1, 1))
```

we want to call directly, but for the following, we would want to call it via `pairwise_kernels`.

```
def unvectorizable_pairwise_kernel(X, Y=None):
     ...
```

Should we have a rule that if the number of arguments in the kernel callable is exactly 1, then it is assumed to be some optimized kernel callable that is already vectorized and hence should be called directly?

Otoh, if not, then it is better to pass it through `pairwise_kernels`

adam2392 commented:
I addressed the issue above and just made it clear that we assume the following:

```
If the metric is a callable, it will have either one input for ``X``, or two inputs for ``X`` and
    ``Y``. If one input is passed in, it is assumed that the kernel operates on the entire array to compute
    the kernel array. If two inputs are passed in, then it is assumed that the kernel operates on
    pairwise row vectors from each input array. This callable is parallelized across rows of the input
    using :func:`~sklearn.metrics.pairwise.pairwise_kernels`. Note that ``(X, Y=None)`` is a valid input
    signature for the kernel function and would then get passed to the pairwise kernel function. If
    a callable is passed in, it is generally faster and more efficient if one can define a vectorized
    operation that operates on the whole array at once. Otherwise, the pairwise kernel function will
    call the function for each combination of rows in the input arrays.
```

adam2392 commented on pywhy_stats/independence/kci.py:
```diff
@@ -228,19 +228,19 @@ def _kernel_test(
-        X = np.concatenate((X, Z), axis=1)
-    Kx = compute_kernel(X, metric=kernel_X, centered=centered, n_jobs=n_jobs)
-    Ky = compute_kernel(Y, metric=kernel_Y, centered=centered, n_jobs=n_jobs)
+        # Equivalent to concatenating and then estimating the kernel matrix
+        Kx *= Kz
```
Idk why but this introduces a regression in the unit-tests.

@bloebp

adam2392 commented on pywhy_stats/independence/kci.py:
```diff
@@ -153,45 +219,29 @@ def _kernel_test(
-    Ky = kernel_Y(Y)
+    X, Y, Z = _preprocess_kernel_data(X, Y, Z, normalize_data)
+
+    # if Z is not None:
+    # concatenate the (X, Z) data to compute the K_xz kernel
+    # X = np.concatenate((X, Z), axis=1)
```
@bloebp if you uncomment these lines, the unit-test does not fail. However, the lines:

```
# Equivalent to concatenating and then estimating the kernel matrix
        Kx *= Kz  # type: ignore
```

makes the unit-test fail. Wanna take a look? I'm not sure why, but I'm also not sure when the multiplication trick is valid?

adam2392 commented on pywhy_stats/independence/kci.py:
```diff
@@ -61,12 +77,31 @@ def ind(
 
         kernel_X = func:`sklearn.metrics.pairwise.pairwise_kernels.polynomial`
 
+    If the kernel is a callable, it will have either one input for ``X``, or two inputs for ``X`` and
+    ``Y``. It is assumed that the kernel operates on the entire array to compute
+    the kernel array, that is, the callable performs a vectorized operation to compute the entire kernel matrix.
+
+    If one has an unvectorizable kernel function, then it is advised to use the callable with the
+    :func:`~sklearn.metrics.pairwise.pairwise_kernels` function, which will parallelize the kernel computation
+    across each row of the data.
```
@bloebp this now describes the policy/behavior of the callable and string input for the `kernel_XYZ` parameter

adam2392 commented:
@bloebp I will let you merge if it looks good to you
